docker/Dockerfile

```diff
@@ -0,0 +1,60 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# syntax=docker/dockerfile:1
+
+FROM registry.access.redhat.com/ubi9/ubi-minimal:latest AS jdk-download
+ARG JDK_DOWNLOAD_LINK
+ARG JDK_VERSION
+ENV JAVA_HOME="/usr/lib/jvm/jdk-${JDK_VERSION}"
+
+RUN \
+    set -xeuo pipefail && \
+    microdnf install -y tar gzip && \
+    # Install JDK from the provided archive link \
+    echo "Downloading JDK from ${JDK_DOWNLOAD_LINK}" && \
+    mkdir -p "${JAVA_HOME}" && \
+    curl -#LfS "${JDK_DOWNLOAD_LINK}" | tar -zx --strip 1 -C "${JAVA_HOME}"
+
+# Use ubi9 minimal as it's more secure
+FROM registry.access.redhat.com/ubi9/ubi-minimal:latest
+WORKDIR /opt/trino
+
+ARG JDK_VERSION
+ENV JAVA_HOME="/usr/lib/jvm/jdk-${JDK_VERSION}"
+ENV PATH=$PATH:$JAVA_HOME/bin
+COPY --from=jdk-download $JAVA_HOME $JAVA_HOME
+
+RUN \
+    set -xeu && \
+    microdnf update -y && \
+    microdnf install -y tar less python3 python3-pip shadow-utils jq && \
+    update-alternatives --install /usr/bin/python python /usr/bin/python3 1 && \
+    pip3 install requests pyyaml && \
+    groupadd trino --gid 1000 && \
+    useradd trino --uid 1000 --gid 1000 --create-home && \
+    mkdir -p /usr/lib/trino && \
+    chown -R "trino:trino" /usr/lib/trino /opt/trino
+
+ARG TRINO_GATEWAY_VERSION
+COPY --chown=trino:trino gateway-ha-${TRINO_GATEWAY_VERSION}-jar-with-dependencies.jar /usr/lib/trino
+COPY --chown=trino:trino gateway-ha /usr/lib/trino
+
+ENV GATEWAY_JAR_PATH "/usr/lib/trino/gateway-ha-${TRINO_GATEWAY_VERSION}-jar-with-dependencies.jar"
+
+EXPOSE 8080
+USER trino:trino
+CMD java --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED -jar "${GATEWAY_JAR_PATH}" "server" "/opt/trino/gateway-ha-config.yml"
```

Comment:
https://github.com/airlift/airlift/tree/master/launcher could be used here

Reply:
Thanks for the quick review 😄 

I began looking at the Airlift launcher but would that be something we should be doing once we migrate the Gateway off of Dropwizard? Or is there a simple way to wrap the existing web server that I am missing?

Will try to find more time to dive into the launcher setup within the next week

Reply:
I think we might delay using launcher .. the problem is also that it pull in a requirement for Python .. ideally we could avoid that. For now I think we are okay with just using the command.

Reply:
@mosabua valid point on the Python requirements, was mostly looking for a non-hacky solution for pulling configurations from YAML in the health checker

I also considered installing jq+yq then doing it in bash, open to other solutions

would wrapping the installs in a venv be sufficient? or would we prefer to limit pip installs completely for Python in this image

Reply:
To loop back here, the Trino health checking script is a bit simpler because extraction from a .properties file is pretty straight forward

In our case the Gateway is leveraging YAML which complicates things slightly, that's why I was hoping to use a qualified YAML parsing package in Python.

I can think of a couple of options if we'd like to skip out on Python:

1. Write something with awk, an example I've tested is:
```bash
function get_2nd_level_yaml_key() {
    local parent_key=$1
    local target_key=$2
    local yaml_file_path=$3

    # In order to get the nested key we will leverage awk to do some parsing
    # First we search for the parent_key, once we find it we set 'in_block' to true
    # Then we will attempt to find the 2nd level target key based on:
    #  1. There's only spaces before the target_key
    #  2. There are no '#'s before the target_key indicating it's in a comment
    # If we determine that we have left the 'parent_key' block we will set 'in_block' to false
    awk -v parent_key=${parent_key} \
        -v target_key=${target_key} \
        '
        match($0, "^" parent_key ":") { in_block=1; next } \
        in_block {
            if (match($0, "^[[:space:]]+" target_key ":") \
              && !/^[[:space:]]*#/ ) {
               sub(/:/, "", $2); print $2; exit
            } else if ($0 ~ /^[^\t ]/) {
                in_block=0
            }
        }
        ' \
        ${yaml_file_path}
}
```
2. Install jq + yq (note: jq you can microdnf install, yq we'd need to explicitly download)

The awk version works but would break in the future if we ever nested the keys we were attempting to match inside of requestRouter. Example:
```
requestRouter:
    newKey:
        port: 8081 # this gets matched
    port: 8080
```

I'm open to suggestions on best way to extract the proper ports and scheme

Reply:
If we go along the 'awk' path, to handle the case above I've modified the awk to attempt to find the indentation required to be a 2nd level key:
```bash
function get_2nd_level_yaml_key() {
    local parent_key=$1
    local target_key=$2
    local yaml_file_path=$3

    # In order to get the nested key we will leverage awk to do some parsing
    # First we search for the parent_key, once we find it we set 'in_block' to true
    # Next because we don't know the indentation levels being provided, we will attempt
    # to find the indentation level of the 2nd level keys.
    # Then we will attempt to find the 2nd level target key based on:
    #  1. There's only spaces before the target_key
    #  2. The indentation level is equal to the 'found' 2nd level indentation
    #  3. There are no '#'s before the target_key indicating it's in a comment
    # If we decide that the key matches our expectations, we will print the matched value
    # If we determine that we have left the 'parent_key' block we will set 'in_block' to false
    awk -v parent_key=${parent_key} \
        -v target_key=${target_key} \
        '
        function count_indentation(line) {
            match(line, /^[[:space:]]*/);
            return RLENGTH;
        }
        match($0, "^" parent_key ":") {
            in_block=1;
            parent_indent_level = count_indentation($0);
            next;
        } \
        in_block {
            current_indent_level = count_indentation($0);
            if ( !first_level_indentation \
              && match($0, "^[[:space:]]+.*:") \
              && current_indent_level > parent_indent_level \
              && !/^[[:space:]]*#/ ) {
                first_level_indentation = current_indent_level;
            }
            if (match($0, "^[[:space:]]+" target_key ":") \
              && current_indent_level == first_level_indentation \
              && !/^[[:space:]]*#/ ) {
               sub(/:/, "", $2);
               print $2;
            } else if ($0 ~ /^[^\t  #]/) {
                in_block=0
            }
        }
        ' \
        ${yaml_file_path}
}
```

Sort of unwieldy 😄

Reply:
Cant we just hit a HTTP endpoint for healthcheck .. also maybe I am missing the point .. but how are healthcheck and launcher related?

Reply:
They're unrelated we could probably spin up a new thread to talk about the health checker, I was replying in this thread as it answered the question of 'why Python'.

> Cant we just hit a HTTP endpoint for healthcheck
The overkill above was mostly just to pull out 2 fields from YAML: `requestRouter.port` and `requestRouter.ssl`. With those 2 parameters we can decide on the proper endpoint to hit for health checking, ex:
```
# Where scheme is http or https
endpoint="${scheme}://localhost:${port}/api/public/backends"
```

Without code changes if we were to roll this out for just HTTP, users would need to override the health checker when using `ssl: true` to use a test healthiness on HTTPS.

TBH there should be dedicated liveness/readiness routes being served by the Gateway itself that we can use, I propose we create a new Github issue for it to minimize the scope of this MR to just docker + release. Thoughts?

(If I'm making any assumptions anywhere please correct me, I'm quite new to this repo 😄 )